### PR TITLE
i#3823 multi-phase drreg: Remove slot id labels.

### PR DESCRIPTION
**Pull request by abhinav92003**
Modifies free spill slot selection logic to use is_our_spill_or_restore instead
of the labels with spill slot use information added for this purpose. We do not
need the extra information in the latter and can simply use the former routine.

Removes the extra spill slot use metadata added in form of labels from the
instrlist.

Also adds documentation about possibility of DR slot conflicts if DR APIs are
mixed with drreg ones.

Issue: #3823

### CODE REVIEW

**Comment by abhinav92003** [comment]
AArch64 failure for `code_api|tool.drcachesim.threads-with-config-file` seems flaky; works fine locally. Will re-run.

**Comment by abhinav92003** [comment]
run arm tests

File: ext/drreg/drreg.dox [hunk 1]
```diff
@@ -53,6 +53,11 @@ less available in future phases; the current logic skips over a slot if
 there's a usage found anywhere later in the bb added by any previous phase.
 So it requires additional spill slots as well.
 
+Note that DR provides APIs to directly access TLS slots: like
+\p dr_insert_write_raw_tls(), \p dr_insert_write_tls_field(). These are
```
Comment (derekbruening) [inline ext/drreg/drreg.dox:57]: This is incorrect: dr_insert_write_tls_field uses the single client field which is unrelated to drreg (except via drmgr multiplexing it to provide drmgr_register_tls_field).

Comment (derekbruening) [reply ext/drreg/drreg.dox:57]: dr_insert_write_raw_tls is perfectly safe to use, and many clients do use it, for slot ranges reserved by dr_raw_tls_calloc.  Only if someone used a slot range they did not actually reserve, which doesn't seem like we need to call it out here.

File: ext/drreg/drreg.dox [hunk 2]
```diff
@@ -53,6 +53,11 @@ less available in future phases; the current logic skips over a slot if
...
+also used by drreg internally. It is unsafe to mix usages of drreg and
+these direct APIs as it may cause conflicts in slot usages.
+
 CAUTION: this extension is still a work in progress.  Not all pieces are
```
Comment (derekbruening) [inline ext/drreg/drreg.dox:61]: I think we can remove this.

File: ext/drreg/drreg.dox (hunk 1 above)
Comment (derekbruening) [reply ext/drreg/drreg.dox:57]: The problem is if drreg is not given enough of its own slots via `#drreg_options_t.num_spill_slots` and so ends up using DR spill slots, and a client simultaneously directly uses DR's spill slots via dr_save_reg()/dr_restore_reg() (and the corresponding dr_read_saved_reg()/dr_write_saved_reg()).

File: ext/drreg/drreg.h [hunk 3]
```diff
@@ -117,9 +117,12 @@ typedef struct _drreg_options_t {
      * requested from DR via dr_raw_tls_calloc().  Any slots needed beyond
      * this number will use DR's base slots, which are not allowed to be
      * used across application instructions.  DR's slots are also more
-     * expensive to access (beyond the first few).
-     * This number should be computed as one plus the number of
-     * simultaneously used general-purpose register spill slots, as
+     * expensive to access (beyond the first few). DR's base slots may
+     * also be used by core DR or other clients, which may cause
```
Comment (derekbruening) [inline ext/drreg/drreg.h:121]: Core DR can be removed here: that's the source of the "not allowed to be used across application instructions"; beyond that there will be no conflict.

Comment (derekbruening) [reply ext/drreg/drreg.h:121]: For "other clients": I would stress that the problem is when using a client or library that doesn't use drreg and instead directly using DR register spill slots (dr_save_reg, etc.); i.e., all cooperating client components should be using drreg.

Comment (abhinav92003) [reply ext/drreg/drreg.h:121]: Done.

File: ext/drreg/drreg.dox (hunk 1 above)
Comment (abhinav92003) [reply ext/drreg/drreg.dox:57]: Thanks. Done.

File: ext/drreg/drreg.dox (hunk 2 above)
Comment (abhinav92003) [reply ext/drreg/drreg.dox:61]: Done.

**Review by derekbruening — APPROVED** [review]

**Pull request merged**